ic-os/components/conformance/check_components.py

```diff
@@ -0,0 +1,119 @@
+#!/usr/bin/env python3
+
+import argparse
+import atexit
+import itertools
+import os
+import pathlib
+import re
+import subprocess
+import sys
+import tarfile
+import tempfile
+
+ALLOWED_UNDECLARED_DEPENDENCIES = {
+    "ic-os/components/misc/fetch-property.sh": {
+        # fetch-property.sh checks existence of metrics.sh
+        "/opt/ic/bin/metrics.sh",
+    }
+}
+
+# Check file patterns /opt/ic/...
+COMPONENT_FILE_PATTERN = r"/opt/ic/[^\s'\"},)]+"
+# Exclude patterns that include $ since we don't support substitution.
+EXCLUDE_FILE_PATTERN = r".*[$].*"
+
+
+def check_paths_in_source(source: str, partition_img_path: str) -> [str]:
+    """
+    Checks if `source` contains only file paths that are present in the image
+    file.
+
+    Args:
+        source: Path to the file to be checked.
+        partition_img_path: Path to the extracted partition file (partition.img)
+
+    Returns a list of errors or empty list on no error.
+    """
+
+    print(f"Checking component {source}")
+
+    if not os.path.isfile(source):
+        return [f"File '{source}' is not a valid file."]
+
+    errors = []
+    allowlisted_unavailable_dependencies = ALLOWED_UNDECLARED_DEPENDENCIES.get(
+        source, {})
+    source_content = pathlib.Path(source).read_text()
+    for dependency in allowlisted_unavailable_dependencies:
+        if dependency not in source_content:
+            errors.append(f"{dependency} is in the allowlist of {source} even "
+                          f"though {source} does not contain any references to "
+                          f"it. Remove it from ALLOWED_UNDECLARED_DEPENDENCIES")
+
+    for line_number, line in enumerate(source_content.splitlines(), start=1):
+        dependencies = re.findall(COMPONENT_FILE_PATTERN, line)
+        for dependency in dependencies:
+            if re.match(EXCLUDE_FILE_PATTERN, dependency):
+                print(f"Skipping dependency {dependency} (unable to parse)")
```

Comment:
This comes across as if the file could not be properly read, maybe "unable to parse path"?